daphne_worker/src/roles/aggregator.rs

```diff
@@ -48,133 +46,112 @@ impl DapReportInitializer for DaphneWorker<'_> {
+                            DURABLE_AGGREGATE_STORE_CHECK_COLLECTED,
+                            durable_name,
+                        )
+                        .map_ok(move |collected| (bucket, collected))
+                })
+                .buffer_unordered(usize::MAX)
+                .try_collect::<HashMap<&DapBatchBucket, bool>>()
+                .map_err(|e| fatal_error!(err = ?e, "failed to check collected"))
+                .await?
+        };
 
-        // Flatten the responses from ReportsProcessed into a hash map.
-        let mut initialized_reports = HashMap::new();
-        for reports_processed_response in reports_processed_responses.into_iter() {
-            for initialized_report in reports_processed_response.initialized_reports.into_iter() {
-                initialized_reports
-                    .insert(initialized_report.metadata().id.clone(), initialized_report);
-            }
-        }
+        let min_time = self.least_valid_report_time(self.get_current_time());
+        let max_time = self.greatest_valid_report_time(self.get_current_time());
+
+        let mut initialized_reports = consumed_reports
+            .into_iter()
+            .map(|consumed_report| {
+                let metadata = consumed_report.metadata();
+
+                let initialized = if metadata.time < min_time {
+                    consumed_report
+                        .into_initialized_rejected_due_to(TransitionFailure::ReportDropped)
+                } else if metadata.time > max_time {
+                    consumed_report
+                        .into_initialized_rejected_due_to(TransitionFailure::ReportTooEarly)
+                } else if collected_reports
+                    [&task_config.bucket_for(part_batch_sel, &consumed_report)]
+                {
+                    consumed_report
+                        .into_initialized_rejected_due_to(TransitionFailure::BatchCollected)
+                } else {
+                    EarlyReportStateInitialized::initialize(
+                        is_leader,
+                        &task_config.vdaf_verify_key,
+                        &task_config.vdaf,
+                        consumed_report,
+                    )?
+                };
 
-        // Send AggregateStore requests.
-        let mut agg_store_requests = Vec::new();
-        for durable_name in agg_store_request_name {
-            agg_store_requests.push(durable.get(
-                BINDING_DAP_AGGREGATE_STORE,
-                DURABLE_AGGREGATE_STORE_CHECK_COLLECTED,
-                durable_name,
-            ));
-        }
-        let agg_store_responses: Vec<bool> = try_join_all(agg_store_requests)
+                Ok(initialized)
+            })
+            .collect::<Result<Vec<_>, DapError>>()
+            .map_err(|e| fatal_error!(err = ?e, "failed to initialize a report"))?;
+
+        let replayed_reports_check = futures::stream::iter(reports_processed_request_data)
```

Comment:
Since replay protection has now moved to AggregateStore, I don't think there's any sense in doing replay protection here as well. We will eventually delete this code --- any reason not to do so now?

Reply:
Good catch. I suppose this will get simpler when we move the collected check to AggregateStore as well?